bpf: add hairpin flow test

dylandreimerink opened the pull request:
This works adds an integration test that the hairpin flow works as
expected(part of #20160). A hairpin flow happens when a pod attempts to call its own
service, causing the packet to go back to that same pod.

While writing the test, a few improvements were made to the test tooling
which is also included. This includes a packet builder mini-library to make
crafting packets in eBPF simpler. Also the ability to dump the context after 
a test has been ran to inspect the actual packet manually. Also made the 
execution order of tests within the same file predictable so multiple tests
can run one after the other on the same map state, which is used in the 
hairflow test to test the return flow after connection tracking entries have 
been created.

borkmann reviewed (approved):
🚀

brb reviewed (approved):
LGTM, thanks! I really like the code comments you added - makes it easy to follow the code.

Minor nit: you could move the pktgen to a separate commit msg.

NikAleksandrov commented on bpf/tests/hairpin_flow.c:
```diff
@@ -0,0 +1,305 @@
+/* Enable redirecting from veth to veth within the host */
+#define ENABLE_HOST_REDIRECT
+
+/* Skip ingress policy checks, not needed to validate hairflow */
```
should hairflow be hairpin flow here?

NikAleksandrov commented on bpf/tests/pktgen.h:
```diff
@@ -0,0 +1,507 @@
+#define default_data "Should not change!!"
+
+enum pkt_layer {
+	PKT_LYR_NONE,
```
minor nit: why not name it just PKT_LAYER as the enum?

dylandreimerink replied:
Well, I would say `PACKET_LAYER` or `PKT_LYR`, why shorten packet and not layer?

NikAleksandrov replied:
I don't care about packet, lyr just looks weird. If you'd like to make it PACKET_LAYER, go for it.
Not indicative, just ran them out of curiousity:
$ grep --include=* -R LYR * | wc -l
1
cilium $ grep --include=* -R LAYER | wc -l
91
cilium $ grep --include=* -R PKT | wc -l
403
cilium ]$ grep --include=* -R PACKET | wc -l
290

dylandreimerink replied:
Changed it from LYR to LAYER

NikAleksandrov commented on bpf/tests/hairpin_flow.c:
```diff
@@ -0,0 +1,305 @@
+	__u32 *status_code = data;
+
+	if (*status_code != TC_ACT_REDIRECT) {
+		test_log("status code = %lld", *status_code);
```
minor nit again: %u instead?

dylandreimerink replied:
I believe `%ul` (unsigned long) which is a `u32` I believe.

NikAleksandrov replied:
unsinged long is 64 bits on Linux/x86_64 for both gcc and clang.

NikAleksandrov commented on bpf/tests/hairpin_flow.c:
```diff
@@ -0,0 +1,305 @@
+
+#include "common.h"
+
+/* Set the LXC source add rest to the address of pod one */
```
I don't quite follow this comment, should "add rest" be "address" ?

NikAleksandrov reviewed (changes requested):
Could you please follow the current bpf code style (which is similar to the kernel's code style) and define the variables in the beginning of functions (or code blocks) that use them consistently? It's harder for me to follow the code with the variables defined everywhere.

Also minor nit: there are a few multiline comment styles mixed, I'd suggest to stick to one.
For example kernel networking subsystem uses:
```
/* line1
 * line2
 */
```

ldelossa commented on bpf/node_config.h:
```diff
@@ -13,14 +13,22 @@
  */
 #include "lib/utils.h"
 
+#ifndef NODE_MAC
```
these strictly necessary or are you being defensive?

dylandreimerink replied:
It is actually necessary. We "redefine" this macro in the new [config_replacement.h](https://github.com/dylandreimerink/cilium/blob/c8947a3a061a6b978daace6fe34afcbcc7d4b947/bpf/tests/config_replacement.h#L76). This is necessary since the test framework can't do global data inlineing yet.

ldelossa commented on bpf/tests/pktgen.h:
```diff
@@ -0,0 +1,507 @@
+ *  having to come up with custom ips.
+ */
+
+#define IPv4(a, b, c, d) (((d) << 24) + ((c) << 16) + ((b) << 8) + (a))
```
Why this and not the use of something like `inet_pton` and the standard in*_addr structs?

dylandreimerink replied:
We can't use `inet_pton`, it is not available in eBPF, even if it was, it takes addresses as `char*` so I am not convinced clang would be able to inline that function if we were to for example copy it. As for the `in*_addr` structs, this might be useful when we add IPv6, but the IPv4 version (`in_addr`) just [contains](https://github.com/openbsd/src/blob/master/include/arpa/inet.h#L110) a single 32-bit field called `s_addr` so I don't feel like it adds much.

ldelossa reviewed (approved):
LGTM

NikAleksandrov commented on bpf/tests/pktgen.h:
```diff
@@ -0,0 +1,515 @@
+			}
+
+			__u16 v4len = (__be16)(builder->cur_off - builder->layer_offsets[i]);
+			/* Calculate total length, which is IPv4 hdr + all layers after it */
```
please move this definition with the rest or add a block to this case

NikAleksandrov commented on bpf/tests/pktgen.h:
```diff
@@ -0,0 +1,515 @@
+				break;
+			}
+
+			__be16 v6len = (__be16)(builder->cur_off + sizeof(struct ipv6hdr) -
```
same as v4len above

NikAleksandrov commented on bpf/tests/pktgen.h:
```diff
@@ -0,0 +1,515 @@
+			 * and start of data in 32-bit words (bytes divided by 4).
+			 */
+
+			__u16 hdr_size;
```
same as v4len above

NikAleksandrov commented on bpf/tests/hairpin_flow.c:
```diff
@@ -0,0 +1,318 @@
+	status_code = data;
+
+	if (*status_code != TC_ACT_REDIRECT) {
+		test_log("status code = %lu", *status_code);
```
As I mentioned in the previous review unsigned long is 64 bits on linux/x86_64. It will get promoted so not a big deal, but just noting

dylandreimerink replied:
Technically, we have our own interpretation of the placeholder values which are platform independent. https://github.com/cilium/cilium/blob/master/test/bpf_tests/bpf_test.go#L423. I have simply removed this log for now since it didn't add much value anyway.

NikAleksandrov commented on bpf/tests/hairpin_flow.c:
```diff
@@ -0,0 +1,318 @@
+	status_code = data;
+
+	if (*status_code != TC_ACT_REDIRECT) {
+		test_log("status code = %lld", *status_code);
```
status_code is u32, "%u" or to be consistent with the above "%lu"

NikAleksandrov commented on bpf/tests/hairpin_flow.c:
```diff
@@ -0,0 +1,318 @@
+	l3 = data + sizeof(__u32) + sizeof(struct ethhdr);
+
+	if ((void *)l3 + sizeof(struct iphdr) > data_end)
+		test_fatal("pkt + data out of bounds");
```
does test_fatal log the place where it failed (e.g. file and line number)?
Asking because there are a few identical error msgs and I'm wondering if we can
differentiate which one we've hit.

dylandreimerink replied:
Updated it to be unique messages, since only the `assert` includes filename + line number info, not the `test_fatal` messages.

NikAleksandrov commented on bpf/tests/hairpin_flow.c:
```diff
@@ -0,0 +1,318 @@
+	l4 = (void *)l3 + sizeof(struct iphdr);
+
+	if ((void *)l4 + sizeof(struct tcphdr) > data_end)
+		test_fatal("pkt + data out of bounds");
```
... e.g. here as well we have the same error msg as above but for l4

nbusseneau reviewed (approved):
Not qualified to discuss BPF changes but test logic LGTM, thanks.

dylandreimerink committed:
bpf: Add packet generation helper for bpf tests

This commit adds a number of variables and functions to make packet
crafting easier. All integration tests(calling into tailcalls) require
the test writer to craft a packet in the packet buffer(xdp_md or
sk_buff). Crafting packets can be tedious because you need to figure out
what "sane" defaults to use(TTL, IP version, IHL), setting correct
length and next-protocol fields ect.

To ease the process, this commit introduces a builder pattern which
can be used to create any arbitrary stack of headers and data. To start
a user defines a `struct pktgen` and links it to the packet buffer with
`pktgen__init()`. The user can then use functions like
`pktgen__push_default_iphdr()` to push a header onto the packet buffer,
an IPv4 header in this case. The function will adjust the tailroom in
the buffer, perform bounds checks, write sane default values and return
a pointer with the proper type. Only a null check is required to
satisfy the verifier, after which you can modify the default header in
any way. After all layers are in place, the user calls
`pktgen__finish()` on the builder, this will perform length calculations
, set the proper next header fields and possibly in the future,
calculate checksums.

This library also provides a set of IPv4 addresses, MAC addresses, and
TCP ports which are made to be distinct and unique. It allows tests
to just pick a number pre-defined address instead of having to come up
with unique addresses and define them separately in each test.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bpf: Add hairpin flow test

This commit add an integration test for the hairpin flow. It tests that
pods can connect to their own services, both in the forward and reverse
flow.

This commit also include a few changes which were necessary to get
TC integration tests to work and some quality of life features.

The modifications to `bpf/ep_config.h` and `bpf/node_config.h` enable
us to overwrite the defined values, necessary since these utilize the
global data inlining macros, which the test framework currently can't
handle. The `bpf/tests/config_replacement.h` file overwrites these
global data variables with the same defaults, just defined statically.

The test framework has been updated to now guarantee that programs
within the same file are ran in alphabetic order of the test name. This
allows us to write multi stage tests (sending 2 packets in this case)
while keeping the map state in between.

The `--dump-ctx` flag has been added to the test framework to
dump the raw context to inspect the crafted packet. And lastly the
makefile has been updated to pass the contents of the `V` param to
the `-v` flag of the test framework. This allows us to run single tests
using a regex that matches on the test name.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>
